x-pack/plugins/fleet/server/routes/data_streams/handlers.ts

```diff
@@ -4,157 +4,183 @@
-        const enhancedDashboards = await getEnhancedDashboards(
-          context.core.savedObjects.client,
-          dashboards
-        );
-
-        packageMetadata[pkg] = {
-          version: pkgSavedObject[0].attributes?.version || '',
-          dashboards: enhancedDashboards,
-        };
+      // Set values from backing indices query
+      dataStreamResponse.dataset = dataset.buckets[0]?.key || '';
+      dataStreamResponse.namespace = namespace.buckets[0]?.key || '';
+      dataStreamResponse.type = type.buckets[0]?.key || '';
+
+      // Find package saved object
+      const pkgName = dataStreamResponse.package;
+      const pkgSavedObject = pkgName ? packageSavedObjectsByName[pkgName] : null;
+
+      if (pkgSavedObject) {
+        // if
+        // - the data stream is associated with a package
+        // - and the package has been installed through EPM
+        // - and we didn't pick the metadata in an earlier iteration of this map()
+        if (!packageMetadata[pkgName]) {
+          // then pick the dashboards from the package saved object
+          const dashboards =
+            pkgSavedObject.attributes?.installed_kibana?.filter(
+              (o) => o.type === KibanaSavedObjectType.dashboard
+            ) || [];
+          // and then pick the human-readable titles from the dashboard saved objects
+          const enhancedDashboards = await getEnhancedDashboards(
+            context.core.savedObjects.client,
+            dashboards
+          );
+
+          packageMetadata[pkgName] = {
+            version: pkgSavedObject.attributes?.version || '',
+            dashboards: enhancedDashboards,
+          };
+        }
+
+        // Set values from package information
+        dataStreamResponse.package = pkgName;
+        dataStreamResponse.package_version = packageMetadata[pkgName].version;
+        dataStreamResponse.dashboards = packageMetadata[pkgName].dashboards;
       }
 
-      return {
-        index: indexName,
-        dataset: datasetBuckets.length ? datasetBuckets[0].key : '',
-        namespace: namespaceBuckets.length ? namespaceBuckets[0].key : '',
-        type: typeBuckets.length ? typeBuckets[0].key : '',
-        package: pkgSavedObject.length ? pkg : '',
-        package_version: packageMetadata[pkg] ? packageMetadata[pkg].version : '',
-        last_activity: lastActivity,
-        size_in_bytes: indexStats[indexName] ? indexStats[indexName].total.store.size_in_bytes : 0,
-        dashboards: packageMetadata[pkg] ? packageMetadata[pkg].dashboards : [],
-      };
+      return dataStreamResponse;
     });
 
-    const dataStreams: DataStream[] = await Promise.all(dataStreamsPromises);
-
-    body.data_streams = dataStreams;
-
+    // Return final data streams objects sorted by last activity, decending
+    // After filtering out data streams that are missing dataset/namespace/type fields
+    body.data_streams = (await Promise.all(dataStreamPromises))
```

Comment:
This handler has more logic/statements than most others. What can we move to the service layer?

I think we can move lines 53-177 into a service allowing something like

```suggestion
    body.data_streams = await service.method()
```

Perhaps it's more than that, but I think we should move as much of this code as possible into a service.

Reply:
When this code was written originally, it was an intentional decision to not create a data stream service yet and just leave the code in the handler because 1) there's only 1 data stream API (list), 2) no other place in the plugins needs access to this logic, and 3) no downstream plugins needs access either. I'd like to keep this simple for now until one of these situations change. Is this blocking for you?

Reply:
No, it's not a blocker. Service or not, I think that code should be pulled into functions to keep the handler easier to read/test, but that can be done later